codalab/client/json_api_client.py

```diff
@@ -0,0 +1,299 @@
+from collections import defaultdict
+import urllib2
+
+from codalab.client.bundle_client import BundleClient
+from codalab.client.rest_client import RestClient
+from codalab.common import http_error_to_exception
+
+
+# TODO(sckoo): deal with more error cases
+def wrap_exception(f):
+    def wrapper(*args, **kwargs):
+        try:
+            return f(*args, **kwargs)
+        except urllib2.HTTPError as err:
+            raise http_error_to_exception(err.code, err.read())
+    return wrapper
+
+
+def unpack_document(document):
+    """
+    Unpack a JSON API document into a plain dict, with the relationship keys
+    wired up to the 'included' resources. Supports multiple levels of
+    relationships. Relationships that do not have a corresponding object in the
+    'included' objects will be filled in with a dict that only has the 'id'
+    key. For example:
+
+     >> unpack_document({
+            'data': {
+                'id': '123',
+                'type: 'bundles',
+                'attributes': {
+                    'name': 'hello'
+                },
+                'relationships': {
+                    'owner': {
+                        'data': {
+                            'id': '345',
+                            'type': 'users'
+                        }
+                    },
+                    'parent': {
+                        'data': {
+                            'id': '567',
+                            'type': 'bundles'
+                        }
+                    }
+                },
+            },
+            'included': [
+                {
+                    'type': 'users':
+                    'id': '345',
+                    'attributes': {
+                        'name': 'percy',
+                        'affiliation': 'stanford'
+                    }
+                }
+            ]
+        })
+
+     => {
+            'owner': {
+                'id': '345',
+                'name': 'percy',
+                'affiliation: 'stanford'
+            },
+            'parent': {
+                'id': '567'
+            },
+            'id': '123',
+            'name': 'hello'
+        }
+
+    :param document: the JSON-API payload as a dict
+    :return:
+    """
+    def unpack_object(obj_data):
+        obj = obj_data['attributes']
+        obj['id'] = obj_data['id']
+
+        if 'meta' in obj_data:
+            obj['meta'] = obj_data['meta']
+
+        for key, relationship in obj_data.get('relationships', {}).iteritems():
+            if isinstance(relationship['data'], list):
+                obj[key] = []
+                for linkage in relationship['data']:
+                    if linkage['id'] in included[linkage['type']].keys():
```

Comment:
You're relying on defaultdict behavior here... That's kind of weird. I'd use a tuple: type, id